src/main/java/com/fasterxml/jackson/databind/ser/std/CollectionSerializer.java

```diff
@@ -38,22 +48,17 @@ public class CollectionSerializer
     public CollectionSerializer(JavaType elemType, boolean staticTyping, TypeSerializer vts,
             JsonSerializer<Object> valueSerializer) {
         super(Collection.class, elemType, staticTyping, vts, valueSerializer);
-    }
-
-    /**
-     * @deprecated since 2.6
-     */
-    @Deprecated // since 2.6
-    public CollectionSerializer(JavaType elemType, boolean staticTyping, TypeSerializer vts,
-            BeanProperty property, JsonSerializer<Object> valueSerializer) {
-        // note: assumption is 'property' is always passed as null
-        this(elemType, staticTyping, vts, valueSerializer);
+        // Unfortunately we can't check for EnumSet statically (if type indicated it,
+        // we'd have constructed `EnumSetSerializer` instead). But we can check that
+        // element type could possibly be an Enum.
+        _maybeEnumSet = elemType.isEnumType() || elemType.isJavaLangObject();
```

Comment:
For 3.x version, would it be worth the effort trying to find other way instead of this?
And also parts with below code

```
// [databind#4849]/[databind#4214]: need to check for EnumSet
        final TypeSerializer typeSer = (_maybeEnumSet && value instanceof EnumSet<?>)
                ? null : _valueTypeSerializer;
                
```

Reply:
If there is a way, sure, it would be great to get rid of this work-around.
But I am not sure how, considering the problem itself and how it occurs.
That is: I do not know how this could be avoided at all, regardless of compatibility concerns.